docs/proposals/20230525-openstack-microversion-support.md

```diff
@@ -0,0 +1,133 @@
+```text
+  ┌────────────────────────────────────┐
+  │                                    │
+  │                                    │
+  │    Versions supported by           │
+  │    OpenStack instance              │
+  │                                    │
+  │    ┌─────────────────────────┐     │
+  │    │                         │     │
+  │    │  Versions available to  │     │
+  │    │  the user               │     │
+  └────┼─────────────────────────┼─────┘
+       │                         │
+       │  Versions supported     │
+       │  by CAPO                │
+       │                         │
+       └─────────────────────────┘
+```
+
+"Versions" could be changed to "features" in the diagram also, but a version difference does not always expose a difference in features at the CAPO level.
+Therefore "versions" is used in the diagram to highlight the need to support multiple versions even though that may not give any new features.
+
+### Goals
+
+1. Detect user requested features and their version needs.
+
+2. Detect server version support.
+
+### Non-goals
+
+- To allow the user to set the versions freely.
+
+### CAPO microversions history
+
+Initially no microversions were specified (only versions).
+This is still true for all clients except compute.
+It has been bumped from the original v2 in order to add support for specific features:
+
+- 2.53 is needed for server tags (https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/924)
+- 2.60 is needed for multi-attach volumes (https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1498)
+
+## Proposal
+
+### User stories
+
+#### Story 1
+
+As a user of CAPO, I would like to create a cluster in an older OpenStack environment that doesn't support `currently hard coded microversion` and I am fine with limiting the features available to me based on this older version.
+
+#### Story 2
+
+As a user of CAPO, I would like to create a cluster in an OpenStack environment that requires a higher microversion than `currently hard coded microversion`.
+
+#### Story 3
+
+As developer/maintainer of CAPO, I want to ensure that the code is executed (and tested) with a specific microversion so that the result is deterministic.
+
+#### Story 4
+
+As a developer/maintainer of CAPO, I want to keep the code readable and limit the burden of testing and maintaining support for multiple microversions.
+
+### Implementation details and notes
+
+There is already one part of CAPO that actually has dynamic versioning, thanks to Gophercloud.
+The [authentication](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/904381c7f31f8fdda83586f208339cd2aab53e78/pkg/scope/provider.go#L173) flow makes use of [this function](https://github.com/gophercloud/gophercloud/blob/6e1423b668969548d46fd862202a3bf9623b52e0/openstack/client.go#L92-L112) that chooses the version from a list of supported versions, provided by the caller, and the range of versions supported by the server.
```

Comment:
That's not really a microversion. I get the point, but this example is super simple. I believe this is a better example: https://github.com/kubernetes/cloud-provider-openstack/blob/35ce5a0c59da23795fc48b292e3736f8cb0bb10f/pkg/util/openstack/loadbalancer.go#L101-L162

Please note this is not really a microversion, Octavia does regular API versioning and you cannot request an older microversion.

Reply:
Thanks! I added the linked example but I also kept the original since I think it can still be useful for the implementation.